solidity/contracts/integrator/TBTCDepositorProxy.sol

```diff
@@ -0,0 +1,331 @@
+        BitcoinTx.Info calldata fundingTx,
+        Deposit.DepositRevealInfo calldata reveal,
+        bytes32 extraData
+    ) internal returns (uint256) {
+        require(reveal.vault == address(tbtcVault), "Vault address mismatch");
+
+        uint256 depositKey = calculateDepositKey(
+            calculateBitcoinTxHash(fundingTx),
+            reveal.fundingOutputIndex
+        );
+
+        pendingDeposits[depositKey] = true;
+
+        emit DepositInitialized(
+            depositKey,
+            /* solhint-disable-next-line not-rely-on-time */
+            uint32(block.timestamp)
+        );
+
+        // The Bridge does not allow to reveal the same deposit twice and
+        // revealed deposits stay there forever. The transaction will revert
+        // if the deposit has already been revealed so, there is no need to do
+        // an explicit check here.
+        bridge.revealDepositWithExtraData(fundingTx, reveal, extraData);
+
+        return depositKey;
+    }
+
+    /// @notice Finalizes a deposit by calculating the amount of TBTC minted
+    ///         for the deposit and calling the `onDepositFinalized` callback
+    ///         function.
+    /// @param depositKey Deposit key identifying the deposit.
+    /// @dev Requirements:
+    ///      - The deposit must be initialized but not finalized
+    ///        (in the context of this contract) yet.
+    ///      - The deposit must be finalized on the Bridge side. That means the
+    ///        deposit must be either swept or optimistically minted.
+    // slither-disable-next-line dead-code
+    function finalizeDeposit(uint256 depositKey) internal {
+        require(pendingDeposits[depositKey], "Deposit not initialized");
+
+        Deposit.DepositRequest memory deposit = bridge.deposits(depositKey);
+        (, uint64 finalizedAt) = tbtcVault.optimisticMintingRequests(
+            depositKey
+        );
+
+        require(
+            deposit.sweptAt != 0 || finalizedAt != 0,
+            "Deposit not finalized by the bridge"
+        );
+
+        // We can safely delete the deposit from the pending deposits mapping.
+        // This deposit cannot be initialized again because the bridge does not
+        // allow to reveal the same deposit twice. Deleting the deposit from
+        // the mapping will also prevent the finalizeDeposit function from
+        // being called again for the same deposit.
+        // slither-disable-next-line reentrancy-no-eth
+        delete pendingDeposits[depositKey];
+
+        uint256 tbtcAmount = calculateTbtcAmount(
+            deposit.amount,
+            deposit.treasuryFee
+        );
+
+        // slither-disable-next-line reentrancy-events
+        emit DepositFinalized(
+            depositKey,
+            tbtcAmount,
+            /* solhint-disable-next-line not-rely-on-time */
+            uint32(block.timestamp)
+        );
+
+        onDepositFinalized(depositKey, tbtcAmount, deposit.extraData);
```

Comment:
🤔 Since the `finalizeDeposit` function is internal and expected to be overridden in the implementation contract, would it make more sense to return the values instead of passing them in a callback?
With the current approach, we expect the contract to implement both `finalizeDeposit` and `onDepositFinalized`, and they are invoked together.

What if we rename the internal function to `_finalizeDeposit` and return the values:
```suggestion
        return (tbtcAmount, deposit.extraData);
```

the implementation could look like this:
```solidity
    function finalizeDeposit(uint256 depositKey) public override {
        uint256 tbtcAmount, bytes32 extraData = _finalizeDeposit(depositKey);

        (address receiver, uint16 referral) = decodeExtraData(extraData);

        request.amountToStake = tbtcAmount;
        request.receiver = receiver;
        request.referral = referral;
    }
```

Reply:
Actually `finalizeDeposit` is not expected to be overridden (it's not `virtual`) but rather called from a child's external function. I was also thinking about returning the parameters but the solution with `onDepositFinalized` seemed more clear and intuitive to me. I don't have a strong preference though. If you find return params more convenient, I can go that way.